### PR TITLE
MonadLogIO

### PR DESCRIPTION
**Pull request by larskuhtz**
`MonadLogIO` is the class of loggers that allow to extract a log function of type `LogFunctionIO`.

``` haskell
-- | Instances of 'MonadLog' that allow to obtain a 'LogFunctionIO' as plain
-- value. This is helpful when dealing with frameworks that take a logging
-- function in 'IO' as parameter.
--
-- An instance of this class should apply the 'LogLevel', 'LogScope', and
-- 'LogPolicy' at the time when 'logFunIO' is called and not when the returned
-- action is excecuted. If the returned action is excecuted after the logger
-- got released or otherwise invalidated the behavior should match the behavior
-- on a congested logging pipeling accorrding to the log-policy that was in
-- scope when 'logFunIO' was called.
--
-- Even though it can be very convenient, instances of this class must be used
-- with care. The action may contain in its closure a reference to some
-- internal state of the 'MonadLog' instance. Beside of being a source of
-- potential memory leaks, there also is nothing that prevents a programer to
-- call this action outside of the valid scope of the 'MonadLog' instance. In
-- case that the context of the 'MonadLog' instance depends on some state that
-- gets explicitely deallocated this action may have unexectped behavior.
```

### CODE REVIEW

File: src/System/Logger/Types.hs [hunk 1]
```diff
@@ -284,6 +285,107 @@ popLabel = localScope $ \case { [] → []; (_:t) → t }
...
+
+    withLoggerPolicy ∷ LogPolicy → ctx → (ctx → α) → α
+    withLoggerPolicy policy ctx f = f $ ctx & setLoggerPolicy .~ policy
+    {-# INLINE withLoggerPolicy #-}
+
+newtype LoggerCtxT ctx m α = LoggerCtxT { unLoggerCtxT ∷ ReaderT ctx m α }
+    deriving (Functor, Applicative, Monad, MonadIO, MonadTrans, MonadReader ctx, MonadError a, MonadState a, MonadWriter a, MonadBase a, MonadTrace t)
+
+-- This should eventually be defined in Control.Monad.Trace.Class
+instance (Monad m, MonadTrace t m) ⇒ MonadTrace t (ReaderT ctx m) where
```
Comment (jonsterling) [inline src/System/Logger/Types.hs:354]: I would prefer to fix this in the package itself rather than release a version of this package with an orphan instance...

File: src/System/Logger/Types.hs [hunk 2]
```diff
@@ -383,76 +485,29 @@ instance (MonadLog a m, MonadTransControl t, Monad n, n ~ (t m)) ⇒ MonadLog a
...
--- If this doesn't fit your needs you may use a newtype wrapper and define
--- your own instances.
---
-class LoggerCtx ctx msg | ctx → msg where
-    loggerFunIO
-        ∷ (Show msg, Typeable msg, NFData msg)
-        ⇒ ctx
-        → LogFunctionIO msg
+instance (MonadLog a (ReaderT σ m), MonadLogIO a m) ⇒ MonadLogIO a (ReaderT σ m) where
+    logFunIO= lift logFunIO
```
Comment (jonsterling) [inline src/System/Logger/Types.hs]: whitespace

File: src/System/Logger/Types.hs [hunk 3]
```diff
@@ -284,6 +285,107 @@ popLabel = localScope $ \case { [] → []; (_:t) → t }
...
+-- scope when 'logFunIO' was called.
+--
+-- Even though it can be very convenient, instances of this class must be used
+-- with care. The action may contain in its closure a reference to some
+-- internal state of the 'MonadLog' instance. Beside of being a source of
+-- potential memory leaks, there also is nothing that prevents a programer to
+-- call this action outside of the valid scope of the 'MonadLog' instance. In
+-- case that the context of the 'MonadLog' instance depends on some state that
+-- gets explicitely deallocated this action may have unexectped behavior.
+--
```
Comment (jonsterling) [inline src/System/Logger/Types.hs:308]: I actually am a little bit concerned about this... While doing it the explicit way is a little inconvenient, one could at least tell quite easily from the code that it was safe...

Comment (larskuhtz) [inline src/System/Logger/Types.hs:308]: I know it's not nice. It's a bit like `unsafePerformIO` for a monad different than IO. I wonder if the same purpose could be achieved with an instance of `MonadBaseControl`?

[On the other hand what consider safe in other places is not necessarily safer than what we have here. For instance when `withLogger ... \logger -> ...` is used nothing (but best practices and convention) is preventing a programer from using `logger` outside the scope of `withLogger`.]

Comment (jonsterling) [inline src/System/Logger/Types.hs:308]: Yeah... I suppose either approach allows the same mistakes probably. The benefit of the `withLogger` thing is that the intended use case is more clear, but it is a bit more of a pain.

Anyway, I'm fine with it if you really want to add this class. Personally, I don't really think it is necessary, but I don't have a super strong objection.

File: src/System/Logger/Types.hs (hunk 1 above)
Comment (larskuhtz) [inline src/System/Logger/Types.hs:354]: Valid point. I am not completely sure how to do this. This isn't a standard mtl style instance and uses the `MonadBaseControl` instance of `ReaderT`. I am not sure if such an instance would be OK in `hs-trace`.

Comment (jonsterling) [inline src/System/Logger/Types.hs:354]: Oh, interesting... If you do not think we want such an instance in `hs-trace`, then perhaps we should use a newtype rather than `ReaderT` here...

Since I'm still mystified at first glance by most of the `MonadBaseControl` stuff, can you summarize for me why the implementation here uses `liftWith` and `restoreT` rather than a standard mtl-style instance?

**Comment by larskuhtz** [comment]
I put this PR on hold until I get back to work on the project for which I needed this feature.

**Comment by larskuhtz** [comment]
This PR got superseded by the release of 0.2.
